### PR TITLE
Check weather data age and refresh via API if older than 2 hours

### PR DESCRIPTION
**Pull request by IhorYu**
Hey team! 🚀

Just pushed some updates that enhance our weather forecast feature. Now, it checks if the weather info for a city is fresher than two hours before making a new API call. If it's older, we fetch the latest data and update our logs with this spanking new info and a timestamp. This should keep our forecasts up-to-date and reduce unnecessary API calls - saving us some bandwidth and potentially costs!

**Would love your eyes on this for a review and any feedback you might have.**

### CODE REVIEW

File: src/controllers/forecast.mjs [hunk 1]
```diff
@@ -26,14 +41,13 @@ export const getForecastForCity = async (req, res) => {
 
 export const getForecastForAllCities = async (_, res) => {
   try {
-    const citiesData = await getFileData("./data/cities.json");
+    const cities = await getFileData(CITIES_FILE_PATH);
+    cities.forEach(async (city) => await checkUpdate(city));
```
Comment (Mr-Arthur-White) [inline src/controllers/forecast.mjs]: And here we can Promise.all it
```suggestion
    const weatherByCities = await Promise.all(cities.map(city => checkUpdate(city)));
```

File: src/controllers/forecast.mjs [hunk 2]
```diff
@@ -1,23 +1,38 @@
...
+    !cityWeather || new Date(cityWeather.lastUpdate).getTime() < twoHoursAgo;
+
+  if (needUpdate) {
+    const forecast = await getForecastByCity(city);
+    weatherLog[cityName] = {
+      ...forecast,
+    };
+  }
+  await writeFile(WEATHER_LOG_FILE_PATH, weatherLog);
+};
```
Comment (Mr-Arthur-White) [inline src/controllers/forecast.mjs]: I would propose to manage this question on service level
Service responsible to return us the data, no matter how

So, in controller we have only business logic in terms of "I have this request, I know that I need to work with city, I know that I can ask service for a forecast and I don't care how exactly. After that I need to return data that service provides to me"

We can also introduce one more util that would be responsible to check expiration
It should know about time-shift condition (grabbing info about outdate amount of ms from config `const EXPIRATION_IN_MS = 2 * 60 * 60 * 1000`)
We might have a function there (`isExpirationValid`) that doesn't care about data itself, it cares only about input number property and returns true/false

File: src/controllers/forecast.mjs [hunk 3]
```diff
@@ -1,23 +1,38 @@
...
     if (!city) {
       return res.status(404).json({ message: "City not found" });
     }
-
-    const forecast = await getForecastByCity(city);
-    res.status(200).json(forecast);
-  } catch (error) {
-    console.error("Error getting forecast for city:", error);
+    await checkUpdate(city);
+    res.status(200).json((await getFileData(WEATHER_LOG_FILE_PATH))[cityName]);
```
Comment (Mr-Arthur-White) [inline src/controllers/forecast.mjs]: Due to main comment, we wouldn't need to read file one more time
Time saved, resources saved

File: src/controllers/forecast.mjs [hunk 4]
```diff
@@ -26,14 +41,13 @@ export const getForecastForCity = async (req, res) => {
...
+    const cities = await getFileData(CITIES_FILE_PATH);
+    cities.forEach(async (city) => await checkUpdate(city));
 
-    const forecasts = await Promise.all(
-      citiesData.map((city) => getForecastByCity(city))
-    );
-    res.status(200).json(forecasts);
-  } catch (error) {
-    console.error("Error getting forecasts for all cities:", error);
+    res.status(200).json(await getFileData(WEATHER_LOG_FILE_PATH));
```
Comment (Mr-Arthur-White) [inline src/controllers/forecast.mjs]: Due to main comment, we wouldn't need to read file one more time
Time saved, resources saved

File: src/utils/expirationCheck.mjs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { EXPIRATION_IN_MS } from "../../config.mjs";
+
+export const isExpirationValid = async (lastUpdate) => {
+  const now = new Date().getTime();
+  const lastUpdateTime = new Date(lastUpdate).getTime();
+  return now - lastUpdateTime < EXPIRATION_IN_MS;
+};
```
Comment (Mr-Arthur-White) [inline src/utils/expirationCheck.mjs]: Why it's async? 😄

Comment (IhorYu) [reply src/utils/expirationCheck.mjs]: Indeed, got a bit carried away there! 😄

**Review by Mr-Arthur-White — APPROVED** [review]
Machine 🦾

**Comment by Mr-Arthur-White** [comment]
Don't forger to delete branches after
This way we can keep our repository clean
Just a good habit

image

**Comment by IhorYu** [comment]
> Machine 🦾

No, it's actually you who's the machine! 🦾

**Pull request merged**